// @microfleet/core/lib/plugins/http/handlers/hapi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const Bluebird = require("bluebird");
const common_errors_1 = require("common-errors");
const hapi_1 = require("@hapi/hapi");
const attach_1 = require("./router/attach");
const defaultPlugins = [{
        options: {},
        plugin: './plugins/redirect',
    }, {
        options: {},
        plugin: './plugins/state',
    }];
function createHapiServer(config, service) {
    const { handlerConfig } = config.server;
    handlerConfig.server.address = config.server.host || '0.0.0.0';
    handlerConfig.server.port = config.server.port || 3000;
    assert(service.hasPlugin('logger'), 'must include logger plugin');
    const server = service.http = new hapi_1.Server(handlerConfig.server);
    let routerPlugin;
    if (config.router.enabled) {
        routerPlugin = attach_1.default(service, config.router);
    }
    // exposes microfleet inside the server for tighter integrations
    server.decorate('server', 'microfleet', service);
    async function initPlugins() {
        const { list, options } = handlerConfig.plugins;
        const plugins = defaultPlugins.concat(list);
        if (handlerConfig.views) {
            plugins.push({
                options: {},
                plugin: '@hapi/vision',
            });
            plugins.push({
                options: handlerConfig.views,
                plugin: './plugins/views',
            });
        }
        if (routerPlugin !== undefined) {
            plugins.push(routerPlugin);
        }
        const registrations = [];
        for (const pluguinConfiguration of plugins) {
            registrations.push({
                options: pluguinConfiguration.options,
                plugin: typeof pluguinConfiguration.plugin === 'string'
                    ? require(pluguinConfiguration.plugin)
                    : pluguinConfiguration.plugin,
            });
        }
        return server.register(registrations, options);
    }
    async function startServer() {
        if (config.server.attachSocketIO) {
            if (!service.socketIO) {
                return Bluebird.reject(new common_errors_1.NotPermittedError('SocketIO plugin not found'));
            }
            service.socketIO.listen(server.listener, service.config.socketIO.options);
        }
        await initPlugins();
        await server.start();
        service.log.info({ transport: 'http', http: '@hapi/hapi' }, 'listening on http://%s:%s', handlerConfig.server.address, handlerConfig.server.port);
        service.emit('plugin:start:http', server);
        return server;
    }
    async function stopServer() {
        await server.stop();
        service.emit('plugin:stop:http', server);
    }
    return {
        close: stopServer,
        connect: startServer,
    };
}
exports.default = createHapiServer;
