// @microfleet/core/lib/index.js
"use strict";
/**
 * Microservice Abstract Class
 * @module Microfleet
 */
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const Bluebird = require("bluebird");
const EventEmitter = require("eventemitter3");
const is = require("is");
const partial = require("lodash.partial");
const constants = require("./constants");
const defaultOpts = require("./defaults");
const validation_1 = require("@microfleet/validation");
const pluginHealthStatus_1 = require("./utils/pluginHealthStatus");
/**
 * Simple invocation that preserves context.
 * @param fn - Function to invoke.
 */
function invoke(fn) {
    return fn.call(this);
}
const toArray = (x) => Array.isArray(x) ? x : [x];
/**
 * Constants with possilble transport values
 * @memberof Microfleet
 */
exports.ActionTransport = constants.ActionTransport;
/**
 * Constants with connect types to control order of service bootstrap
 * @memberof Microfleet
 */
exports.ConnectorsTypes = constants.ConnectorsTypes;
/**
 * Default priority of connectors during bootstrap
 * @memberof Microfleet
 */
exports.ConnectorsPriority = constants.ConnectorsPriority;
/**
 * Plugin Types
 * @memberof Microfleet
 */
exports.PluginTypes = constants.PluginTypes;
/**
 * Plugin boot priority
 * @memberof Microfleet
 */
exports.PluginsPriority = constants.PluginsPriority;
/**
 * Helper method to enable router extensions.
 * @param name - Pass extension name to require.
 * @returns Extension to router plugin.
 */
exports.routerExtension = (name) => {
    return require(`./plugins/router/extensions/${name}`).default;
};
/**
 * @class Microfleet
 */
class Microfleet extends EventEmitter {
    /**
     * @param [opts={}] - Overrides for configuration.
     * @returns Instance of microservice.
     */
    constructor(opts) {
        super();
        /**
         * Overrides SIG* events and exits cleanly.
         * @returns Resolves when exit sequence has completed.
         */
        this.exit = async () => {
            this.log.info('received close signal...\n closing connections...\n');
            try {
                await this.close().timeout(10000);
            }
            catch (e) {
                process.exit(128);
            }
        };
        /**
         * Notifies about errors when no other listeners are present
         * by throwing them.
         * @param err - Error that was emitted by the service members.
         */
        this.onError = (err) => {
            if (this.listeners('error').length > 1) {
                return;
            }
            throw err;
        };
        // init configuration
        this.config = { ...defaultOpts, ...opts };
        // init migrations
        this.migrators = Object.create(null);
        // init health status checkers
        this[constants.HEALTH_CHECKS_PROPERTY] = [];
        // init plugins
        this.plugins = [];
        this[constants.CONNECTORS_PROPERTY] = Object.create(null);
        this[constants.DESTRUCTORS_PROPERTY] = Object.create(null);

        console.dir(this.config);
        this.initPlugins(this.config);
        // setup error listener
        this.on('error', this.onError);
        // setup hooks
        for (const [eventName, hooks] of Object.entries(this.config.hooks)) {
            for (const hook of toArray(hooks)) {
                this.on(eventName, hook);
            }
        }
        if (this.config.sigterm) {
            this.on('ready', () => {
                process.on('SIGTERM', this.exit);
            });
            this.on('close', () => {
                process.removeListener('SIGTERM', this.exit);
            });
        }
    }
    /**
     * Asyncronously calls event listeners
     * and waits for them to complete.
     * This is a bit odd compared to normal event listeners,
     * but works well for dynamically running async actions and waiting
     * for them to complete.
     *
     * @param event - Hook name to be called during execution.
     * @param args - Arbitrary args to pass to the hooks.
     * @returns Result of invoked hook.
     */
    hook(event, ...args) {
        const listeners = this.listeners(event);
        return Bluebird
            .resolve(listeners)
            .map((listener) => {
            return listener.apply(this, args);
        });
    }
    /**
     * Adds migrators.
     * @param name - Migrator name.
     * @param fn - Migrator function to be invoked.
     * @param args - Arbitrary args to be passed to fn later on.
     */
    addMigrator(name, fn, ...args) {
        // @ts-ignore
        this.migrators[name] = partial(fn, ...args);
    }
    /**
     * Performs migration for a given database or throws if migrator is not present.
     * @param  name - Name of the migration to invoke.
     * @param  args - Extra args to pass to the migrator.
     * @returns Result of the migration.
     */
    migrate(name, ...args) {
        const migrate = this.migrators[name];
        assert(is.fn(migrate), `migrator ${name} not defined`);
        return migrate(...args);
    }
    /**
     * Generic connector for all of the plugins.
     * @returns Walks over registered connectors and emits ready event upon completion.
     */
    connect() {
        return Bluebird
            .resolve(this.processAndEmit(this.getConnectors(), 'ready', exports.ConnectorsPriority));
    }
    /**
     * Generic cleanup function.
     * @returns Walks over registered destructors and emits close event upon completion.
     */
    close() {
        return Bluebird
            .resolve(this.processAndEmit(this.getDestructors(), 'close', [...exports.ConnectorsPriority].reverse()));
    }
    // ****************************** Plugin section: public ************************************
    /**
     * Public function to init plugins.
     *
     * @param mod - Plugin module instance.
     * @param mod.name - Plugin name.
     * @param mod.attach - Plugin attach function.
     * @param [conf] - Configuration in case it's not present in the core configuration object.
     */
    initPlugin(mod, conf) {
        this.plugins.push(mod.name);
        let expose;
        try {
            expose = mod.attach.call(this, conf || this.config[mod.name], __filename);
        }
        catch (e) {
            if (e.constructor === validation_1.HttpStatusError) {
                e.message = `[@microfleet/core] Could not attach ${mod.name}:\n${e.message}`;
            }
            throw e;
        }
        if (!is.object(expose)) {
            return;
        }
        const { connect, status, close } = expose;
        const type = exports.ConnectorsTypes[mod.type];
        assert(type, 'Plugin type must be equal to one of connectors type');
        if (typeof connect === 'function') {
            this.addConnector(type, connect);
        }
        if (typeof close === 'function') {
            this.addDestructor(type, close);
        }
        if (typeof status === 'function') {
            this.addHealthCheck(new pluginHealthStatus_1.PluginHealthCheck(mod.name, status));
        }
    }
    /**
     * Returns registered connectors.
     * @returns Connectors.
     */
    getConnectors() {
        return this[constants.CONNECTORS_PROPERTY];
    }
    /**
     * Returns registered destructors.
     * @returns Destructors.
     */
    getDestructors() {
        return this[constants.DESTRUCTORS_PROPERTY];
    }
    /**
     * Returns registered health checks.
     * @returns Health checks.
     */
    getHealthChecks() {
        return this[constants.HEALTH_CHECKS_PROPERTY];
    }
    /**
     * Initializes connectors on the instance of Microfleet.
     * @param {string} type - Connector type.
     * @param {Function} handler - Plugin connector.
     */
    addConnector(type, handler) {
        this.addHandler(constants.CONNECTORS_PROPERTY, type, handler);
    }
    /**
     * Initializes destructor on the instance of Microfleet.
     * @param {string} type - Destructor type.
     * @param {Function} handler - Plugin destructor.
     */
    addDestructor(type, handler) {
        this.addHandler(constants.DESTRUCTORS_PROPERTY, type, handler);
    }
    /**
     * Initializes plugin health check.
     * @param {Function} handler - Health check function.
     */
    addHealthCheck(handler) {
        this[constants.HEALTH_CHECKS_PROPERTY].push(handler);
    }
    /**
     * Asks for health status of registered plugins if it's possible, logs it and returns summary.
     */
    getHealthStatus() {
        return pluginHealthStatus_1.getHealthStatus(this.getHealthChecks(), this.config.healthChecks);
    }
    hasPlugin(name) {
        return this.plugins.includes(name);
    }
    /**
     * Helper for calling funcs and emitting event after.
     *
     * @param collection - Object with namespaces for arbitrary handlers.
     * @param event - Type of handlers that must be called.
     * @param [priority=Microfleet.ConnectorsPriority] - Order to process collection.
     * @returns Result of the invocation.
     */
    async processAndEmit(collection, event, priority = exports.ConnectorsPriority) {
        const responses = [];
        for (const connectorType of priority) {
            const connectors = collection[connectorType];
            if (!connectors) {
                continue;
            }
            responses.push(...await Bluebird.resolve(connectors).bind(this).map(invoke));
        }
        this.emit(event);
        return responses;
    }
    // ***************************** Plugin section: private **************************************
    addHandler(property, type, handler) {
        if (this[property][type] === undefined) {
            this[property][type] = [];
        }
        this[property][type].push(handler);
    }
    /**
     * Initializes service plugins.
     * @param {Object} config - Service plugins configuration.
     * @private
     */
    initPlugins(config) {
        for (const pluginType of exports.PluginsPriority) {
            this[constants.CONNECTORS_PROPERTY][pluginType] = [];
            this[constants.DESTRUCTORS_PROPERTY][pluginType] = [];
        }
        // require all modules
        const plugins = [];
        for (const plugin of config.plugins) {
            plugins.push(require(`./plugins/${plugin}`));
        }
        // sort and ensure that they are attached based
        // on their priority
        plugins.sort(this.pluginComparator);
        // call the .attach function
        for (const plugin of plugins) {
            this.initPlugin(plugin);
        }
        this.emit('init');
    }
    pluginComparator(a, b) {
        const ap = exports.PluginsPriority.indexOf(a.type);
        const bp = exports.PluginsPriority.indexOf(b.type);
        // same plugin type, check priority
        if (ap === bp) {
            if (a.priority < b.priority)
                return -1;
            if (a.priority > b.priority)
                return 1;
            return 0;
        }
        // different plugin types, sort based on it
        if (ap < bp)
            return -1;
        return 1;
    }
}
constants.CONNECTORS_PROPERTY, constants.DESTRUCTORS_PROPERTY, constants.HEALTH_CHECKS_PROPERTY;
exports.Microfleet = Microfleet;
// if there is no parent module we assume it's called as a binary
if (!module.parent) {
    const mservice = new Microfleet({ name: 'cli' });
    mservice
        .connect()
        .catch((err) => {
        mservice.log.fatal('Failed to start service', err);
        setImmediate(() => { throw err; });
    });
}
