// @microfleet/core/lib/plugins/socketIO/router/adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _debug = require("debug");
const noop = require("lodash/noop");
const __1 = require("../../..");
const debug = _debug('mservice:router:socket.io');
const { socketIO } = __1.ActionTransport;
function getSocketIORouterAdapter(_, router) {
    return function socketIORouterAdapter(socket) {
        socket.on('*', (packet) => {
            const [actionName, params, callback] = packet.data;
            const request = {
                socket,
                params,
                action: noop,
                headers: Object.create(null),
                locals: Object.create(null),
                log: console,
                method: 'socketio',
                parentSpan: undefined,
                query: Object.create(null),
                route: '',
                span: undefined,
                transport: socketIO,
                transportRequest: packet,
            };
            debug('prepared request with', packet.data);
            router.dispatch.call(router, actionName, request, callback);
        });
    };
}
exports.default = getSocketIORouterAdapter;
