// @microfleet/core/lib/plugins/redisCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const Bluebird = require("bluebird");
const _debug = require("debug");
const eventToPromise = require("event-to-promise");
const __1 = require("../");
const require_1 = require("../utils/require");
const migrate_1 = require("./redis/migrate");
const common_errors_1 = require("common-errors");
const utils_1 = require("./redis/utils");
const debug = _debug('mservice:redisCluster');
/**
 * Plugin name
 */
exports.name = 'redis';
/**
 * Plugin type
 */
exports.type = __1.PluginTypes.database;
/**
 * Relative priority inside the same plugin group type
 */
exports.priority = 0;
/**
 * Attaches Redis Cluster plugin.
 * @param  [conf={}] - Configuration for Redis Cluster Connection.
 * @returns Connections and Destructors.
 */
function attach(opts = {}) {
    const service = this;
    const Redis = require_1.default('ioredis');
    const { ERROR_NOT_STARTED, ERROR_ALREADY_STARTED } = require('./redis/constants');
    assert(service.hasPlugin('validator'), new common_errors_1.NotFoundError('validator module must be included'));
    // push out its own bluebird version and configure cancellation
    Redis.Promise = Bluebird.getNewLibraryCopy();
    Redis.Promise.config({
        cancellation: true,
    });
    const { Cluster } = Redis;
    const isClusterStarted = utils_1.isStarted(service, Cluster);
    const conf = service.ifError('redisCluster', opts);
    return {
        /**
         * @returns Opens redis connection.
         */
        async connect() {
            assert(service.redis == null, ERROR_ALREADY_STARTED);
            const instance = new Cluster(conf.hosts, {
                ...conf.options,
                lazyConnect: true,
            });
            if (service.tracer) {
                const applyInstrumentation = require_1.default('opentracing-js-ioredis');
                applyInstrumentation(service.tracer, instance);
            }
            // attach to instance right away
            if (conf.luaScripts) {
                debug('attaching lua');
                utils_1.loadLuaScripts.call(service, conf.luaScripts, instance);
            }
            const $conn = instance.connect();
            const $ready = eventToPromise(instance, 'ready', { ignoreErrors: true });
            await Bluebird.race([$conn, $ready]);
            // cancel either promise
            if ($conn.isPending()) {
                $conn.cancel();
            }
            else {
                $ready.cancel();
            }
            service.addMigrator('redis', migrate_1.default, instance, service);
            service.redis = instance;
            service.emit('plugin:connect:redisCluster', instance);
            return instance;
        },
        /**
         * @returns Returns current status of redis cluster.
         */
        status: utils_1.hasConnection.bind(service, isClusterStarted),
        /**
         * @returns Closes redis connection.
         */
        async close() {
            assert(isClusterStarted(), ERROR_NOT_STARTED);
            await service.redis
                .quit()
                .catchReturn({ message: 'Connection is closed.' }, null);
            service.redis = null;
            service.emit('plugin:close:redisCluster');
        },
    };
}
exports.attach = attach;
